File: firmware/fw_upgrade.c
#include<stdio.h>
#include<unistd.h>
#include<stdlib.h>
#include<string.h>

#include "uci_conf.h"

static int
read_shell_output(char *cmd, char *buf)
{
    FILE    *fp =NULL;
    size_t  readSize = 0;

    fp = popen(cmd, "r");
    if (!fp){
        printf("popen fail\n");
        return (-1);
    }
    readSize = fread(buf, sizeof(char), 1024-1, fp);
    if (readSize == 0){
        pclose(fp);
        printf("output file size 0, please check to output\n");
        return (-1);
    }
    pclose(fp);
    return 0;
}

static int
eval(char *cmd)
{
    int status;
    status = system(cmd);
    
    if (WEXITSTATUS(status)){
        printf("%s result = %d\n",cmd, WEXITSTATUS(status));
        return WEXITSTATUS(status);
    }
    else{
        printf("system call fail\n");
        return(-1);   
    }
}

static int 
tftp_image_download(char *filename, char *serverip)
{
    char cmdbuf[128];
    int result = chdir("/tmp");

    if (result != 0){
        printf("**Directory move failed**\n");
        return (-1);
    }
    sprintf(cmdbuf, "tftp -gr %s %s", filename, serverip);
    printf("%s\n",cmdbuf);
    return 0;//eval(cmdbuf);
}

static int 
nand_flash_erase(char *mtd_part)
{
    char strbuf[64];

    sprintf(strbuf,"flash_erase %s 0 0", mtd_part);
    printf("%s\n",strbuf);
    return 0;//eval(strbuf);
}

static int 
nand_flash_write(char *mtd_part, char *img)
{
    char strbuf[64];

    sprintf(strbuf,"nandwrite %s -p %s", mtd_part, img);
    printf("%s\n",strbuf);
    return 0;//eval(strbuf);    
}

static int 
active_part_get(char *mtd_part)
{
    int part_num;
    char strbuf[64];

    if (read_shell_output("fw_printenv active_part", strbuf) != 0){
        printf("active_part_get error\n");
        return (-1);
    }
    sscanf(strbuf,"active_part=%d", &part_num);

    switch(part_num){
        case 0 :
            part_num = 1;
            strcpy(mtd_part,"/dev/mtd3");
            break;
        case 1 :
            part_num = 0;
            strcpy(mtd_part,"/dev/mtd2");
            break;
        default :
            part_num = -1;
            break;
    }
    return part_num;
}

int main(void)
{
    int  part_num;
    char mtd_part[16];
    char filename[32];
    char serverip[32];
    char type[16];
    char cmdbuf[64];

    config_get("system.upgrade.filename", filename, sizeof(filename));
    config_get("system.upgrade.server", serverip, sizeof(serverip));
    config_get("system.upgrade.type", type, sizeof(type));

    if (tftp_image_download(filename, serverip) != 0){
        printf("download image error\n");
        return (-1);
    }
    if ((part_num = active_part_get(mtd_part)) == -1){
        printf("not found active part \n");
        return (-1);
    }
    if (nand_flash_erase(mtd_part) != 0){
        printf("nand flash erase fail \n");
        return (-1);
    }
    if (nand_flash_write(mtd_part, filename) != 0){
        printf("nand_flash_write fail \n");
        return (-1); 
    }
    sprintf(cmdbuf,"fw_setenv active_part %d", part_num);
    printf("%s\n",cmdbuf);
    /*if (eval(cmdbuf) !=0){
      printf("active_part set error\n");
      return (-1);
      }*/

    return 0;
}


/*
static int 
mtd_part_get(int part_num, char *mtd_part)
{
    switch (part_num) {
        case 0 : strcpy(mtd_part, "/dev/mtd3"); break;
        case 1 : strcpy(mtd_part, "/dev/mtd2"); break;
        default : return (-1);
    }
    return 0;
}*/

/*
    switch(part_num){
        case 0 :
            if (system("fw_setenv active_part 1") !=0){
                printf("active_part set error\n");
                return (-1);
            }
            printf("fw_setenv active_part 1\n");
            break;
        case 1 : 
            if (system("fw_setenv active_part 0") !=0){
                printf("active_part set error\n");
                return (-1);
            }
            printf("fw_setenv active_part 0\n");
            break;
        default : 
            return (-1);
    }
    return 0;*/

File: firmware/fw_upgrade.py
#!/usr/bin/env python

import os
import sys
import subprocess

from env import *
from misc import *

FW_PRINTENV_ACTIVE_PART = 'fw_printenv active_part'
FW_SETENV = 'fw_setenv'
ACTIVE_PART = 'active_part'

UCI_SHOW_CMD = 'uci show '
UCI_GET_CMD = 'uci get '

EQUAL_DELIMITER = '='

TFTP_GET_CMD = 'tftp -g -r '

PATH_DEV_MTD3 = '/dev/mtd3 '
PATH_DEV_MTD2 = '/dev/mtd2 '

FLASH_ERASE = 'flash_erase '
NAND_WRITE = 'nandwrite '

EXIT_OPTION = '0'
RETURN_OPTION = '1'

def subprocess_open(command):
    popen = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
    (stdoutdata, stderrdata) = popen.communicate()
    returncode = popen.poll()
    return stdoutdata, stderrdata, returncode

def simple_error_response(code, error,end_option=EXIT_OPTION):
    print('An invalid result code was returned (error code: ' + str(code) + ')')
    print('Error message : ' + str(error))
    if end_option = RETURN_OPTION:
	return None
    else:
	sys.exit(-1)
    
def get_uci(conf_file, conf_section, conf_option):
    uci_cmd = '.'.join([conf_file, conf_section, conf_option])
    output, error, code = subprocess_open(UCI_GET_CMD + uci_cmd)
    if not code == 0:
	return simple_error_response(code, error, RETURN_OPTION)

    if '\n' in output:
	output = output.replace('\n','')
    return output
    
def get_os_image_from_tftp(file_name, server_addr):
    os.chdir("/tmp")
    print(TFTP_GET_CMD + file_name + ' ' + server_addr)
    output, error, code = subprocess_open(TFTP_GET_CMD + file_name + ' ' + server_addr)
    if not code == 0:
	return simple_error_response(code, error)

class ConfigNand:
    '''
    Write of Erase nand area
    '''
    def __init__(self, part_num):
	if part_num == '0':
	    self.conf_part = PATH_DEV_MTD3
	elif part_num == '1':
	    self.conf_part = PATH_DEV_MTD2
	else:
	    return None

    def erase(self):
	print(FLASH_ERASE + self.conf_part + '0 0')
	#output, error, code = subprocess_open(FLASH_ERASE + self.conf_part + '0 0')
	if not code == 0:
	    return simple_error_response(code,error)

    def write(self,img):
	if os.path.exists(img):
	    print(NAND_WRITE + self.conf_part + '-p ' + img)
	    #output, error, code = subprocess_open(NAND_WRITE + self.conf_part + '-p ' + img)

class ConfigEnv:
    '''
    active_part=0 ==> mtd2 used
    active_part=1 ==> mtd3 used
    '''
    def get_active_part(self):
	output, error, code = subprocess_open(FW_PRINTENV_ACTIVE_PART)
	if not code == 0:
	    return simple_error_response(code, error)
	else:
	    part_num = output.split('=')[1].replace('\n','')
	    return part_num

    def set_active_part(self, part_num):
	'''
	fw_setenv active_part 1 or 0
	'''
	active_part = ' '.join([FW_SETENV, ACTIVE_PART, part_num])
	print(active_part)
	output, error, code = subprocess_open(active_part)
	if not code == 0:
	    return simple_error_response(code, error)


if __name__ == "__main__":

    file_name = get_uci('system', 'upgrade','filename')
    server_ip = get_uci('system', 'upgrade','server')
    upgrade_type = get_uci('system', 'upgrade','type')
   
    if not file_name or server_ip or upgrade_type:
	print('**System upgrade uci information is unusual**')
	sys.exit(-1)

    if upgrade_type == '1':
	tftp_img = get_os_image_from_tftp(file_name, server_ip)
	    
	env_config = ConfigEnv()
	part_num = env_config.get_active_part()

	nand_config = ConfigNand(part_num)
	nand_config.erase()
	nand_config.write(file_name)

	if part_num == '0':
	    result = env_config.set_active_part('1')

	elif part_num == '1':
	    result = env_config.set_active_part('0')

	print('*** Firmware Upgrade finish! ***')
    
    elif upgrade_type == '0':
	print('To-Do...')



